src/QsCompiler/CompilationManager/EditorCompletion.cs

```diff
@@ -0,0 +1,686 @@
+        /// <exception cref="ArgumentNullException">Thrown when any argument is null.</exception>
+        private static int GetTextIndexFromPosition(CodeFragment fragment, Position position)
+        {
+            if (fragment == null)
+                throw new ArgumentNullException(nameof(fragment));
+            if (position == null)
+                throw new ArgumentNullException(nameof(position));
+
+            var relativeLine = position.Line - fragment.GetRange().Start.Line;
+            var lines = Utils.SplitLines(fragment.Text).DefaultIfEmpty("");
+            var relativeCharacter =
+                relativeLine == 0 ? position.Character - fragment.GetRange().Start.Character : position.Character;
+            if (relativeLine < 0 ||
+                relativeLine >= lines.Count() ||
+                relativeCharacter < 0 ||
+                relativeCharacter > lines.ElementAt(relativeLine).Length)
+            {
+                throw new ArgumentException("Position is outside the fragment range", nameof(position));
+            }
+            return lines.Take(relativeLine).Sum(line => line.Length) + relativeCharacter;
+        }
+
+        /// <summary>
+        /// Resolves the namespace alias and returns its full namespace name. If the alias couldn't be resolved, returns
+        /// the alias unchanged.
+        /// </summary>
+        /// <exception cref="ArgumentNullException">Thrown when any argument is null.</exception>
+        private static string ResolveNamespaceAlias(
+            FileContentManager file, CompilationUnit compilation, Position position, string alias)
+        {
+            if (file == null)
+                throw new ArgumentNullException(nameof(file));
+            if (compilation == null)
+                throw new ArgumentNullException(nameof(compilation));
+            if (position == null)
+                throw new ArgumentNullException(nameof(position));
+            if (alias == null)
+                throw new ArgumentNullException(nameof(alias));
+
+            var nsName = file.TryGetNamespaceAt(position);
+            if (nsName == null)
+                return alias;
+            return compilation.GlobalSymbols.TryResolveNamespaceAlias(
+                NonNullable<string>.New(alias), NonNullable<string>.New(nsName), file.FileName) ?? alias;
+        }
+
+        /// <summary>
+        /// Gets the token index and corresponding code fragment at, or the closest before, the given position in the
+        /// file. Returns the relationship between the found fragment and the fragment that would exist directly at the
+        /// given position, which may be different from <see cref="FragmentRelationship.Self"/> if the given position is
+        /// not part of any fragment.
+        /// <para/>
+        /// If <see cref="FragmentRelationship.Missing"/> is returned, then both the token index and code fragment out
```

Comment:
`GetFragmentAtOrBefore` only returns `Missing` if there is no fragment at the current position and no fragments before (so yes I think this would include position (0, 0)). If the given position is not part of a fragment but there's a fragment before the current position, it doesn't return `Missing`. It's kind of confusing because the two fragments `GetFragmentAtOrBefore` is comparing using `FragmentRelationship` is the fragment that it actually found, and the fragment that *would* be at the given position (even if that fragment doesn't exist)…

I think it might be less confusing to replace `FragmentRelationship` with two pieces of information:

1. Did `GetFragmentAtOrBefore` find a fragment exactly at the given position or did it find one before?
2. If it found a fragment before the position, what is its indentation relative to the indentation at the given position?

Reply:
Resolved by refactoring and removing the `FragmentRelationship` enum.